### ISSUE TITLE
Slow. It's just slow :(

### ISSUE THREAD

**Issue by Vittel** [Codewars/codewars.com#1784]
I have had a quick look a the tickets, And didn't see it. Sorry if duplicate.

But why the heck is the page so horribly slow?
I could understand it when executing katas, but on simple things like adding to a list?

There should really be something done about that.

**Comment by kazk** [comment]
We're aware and I'm sorry.

> There should really be something done about that.

I have been working on it.

> why the heck is the page so horribly slow?

The main reason is technical dept including database design that doesn't scale and inefficient code.

I inherited the project late last year and I can say that it's faster and more stable now even with increased users.
I haven't been able to work on performance much because I had to modernize (to some extent) first since many of the dependencies were EOLed. The code was written mostly in mid 2012 to late 2014 and no one really worked on it since.

> I could understand it when executing katas

The service to run the code was replaced last year (#1336) because it was so slow and unstable. It's now often faster to run code than to load some pages.

> on simple things like adding to a list

Can you elaborate on which feature you're talking about?

**Comment by Vittel** [comment]
Adding a kata to a kata collection is what i ment.
Thank you for the explanations.

You could consider some domain driven tweaks to untangle the application step by step and also to make that inefficient database design less important. 
I think of e.g. having your own database for these kata collections. But as i see, this seems to be just the tip of that Eisberg.

I have gained some xp in performance optimization of legacy dotnet pages in the recent time. Let me know if i can help.

**Comment by Hippunky** [comment]
I'd like to second Vittel on the slowness - does performance improve at all on basic navigation of the site for paid users, or is everyone in the same boat? I'm pretty good at SQL optimization, so if you need a hand, I'd be happy to help the optimization efforts.

**Comment by kazk** [comment]
Everyone is in the same boat. Codewars uses MongoDB (using it with relational data is part of the reason why it's so slow).

Closing because it's not actionable.

**Comment by Vittel** [comment]
Well, when u got relational data in mongo, it's already relational, so why not adding a db adapter for a relational db?

Also why are u using relational data with mongo? 

I know how these things come together, but they shouldn't be "wont fix"d that fast. 
Just optimize one call after the other.

**Comment by kazk** [comment]
> I know how these things come together, but they shouldn't be "wont fix"d that fast. Just optimize one call after the other.

I didn't close this because it's a wontfix. I said this issue is not actionable. I answered your question why it's slow. If you meant to open an issue about adding a challenge to a collection, then please open an issue about the specific issue.

> Also why are u using relational data with mongo?

I don't know. I'm just maintaining the inherited project.

---

I'll repost my comment from Gitter explaining why Codewars is in current state.

> CW doesn't even have one dev at the moment.
> I'm the only one that touches the code and I don't work on CW full time. The amount of time I can spend on CW each week depends on how busy I am. So maybe 1/5 dev?
> Other than myself, one of my colleagues from Qualified is helping with GitHub issues and he also does the newsletter.
>
> The reason why CW is sooo slow and have unpredictable performance is not because of Rails. Rails is not known for speed, but it's usually fast enough.
CW is slow because of the inefficient use of the database. 90% of the time is spent working with the database.
> For example, CW uses MongoDB and it has lots of relational data (not a good idea); lots of N+1 queries which is hidden by ORM; deep embedded documents that's written and queried often; some documents are huge and retrieving a set of documents can require multiple requests etc.
>
> I'm assuming it performed fine 5-6 years ago with much less data, but it doesn't scale well. We currently have over 1M users registered and have over 22M documents containing multiple solutions and metadata.
> There's no easy fix for this. Increasing the number of servers won't help. It'll get slower over time and I'm trying to minimize that.
I'm hoping that updating a database version (planned during this quarter) will help with performance a little bit.

> **TLDR** We're really understaffed at the moment because we're focused on Qualified (doing well btw). We know why CW is slow, but we can't do much because of this. Sorry for the frustrating experience, but CW needs Qualified to exist (CW doesn't make money). Codewars is still important to us and we're starting to look for ways to use with Qualified so we can spend more time on it.
>
> One more thing, when reading "understaffed", people usually suggest "why not accept contributions from the community" without thinking. But having 100 people that understands 1% of the codebase contributing won't help at all with the current situation and will make it worse by creating more work. I'm sure many users can contribute small fixes, but we don't need that. If we had the time to review the PRs for small fixes, we can do it ourselves. CW needs fundamental fixes and this needs few people with good understanding of the entire codebase instead, working full time and communicating closely. We also need to do tons of work just to make the code ready for public.

**Comment by Vittel** [comment]
All fine man. Don't take it personal! Ticket is closed.
